dot/build_spec_integration_test.go

```diff
@@ -0,0 +1,150 @@
+	require.NoError(t, err)
+	genesis.TestGenesis.Genesis = genesis.TestFieldsHR
+	require.Equal(t, genesis.TestGenesis.Genesis.Runtime, jGen.Genesis.Runtime)
+	require.Equal(t, expectedChainType, jGen.ChainType)
+	require.Equal(t, expectedProperties, jGen.Properties)
+
+	// confirm raw fields
+	raw, err := bs.ToJSONRaw()
+	require.NoError(t, err)
+	jGenRaw := genesis.Genesis{}
+	err = json.Unmarshal(raw, &jGenRaw)
+	require.NoError(t, err)
+	genesis.TestGenesis.Genesis = genesis.TestFieldsRaw
+	require.Equal(t, genesis.TestGenesis.Genesis.Raw, jGenRaw.Genesis.Raw)
+	require.Equal(t, expectedChainType, jGenRaw.ChainType)
+	require.Equal(t, expectedProperties, jGenRaw.Properties)
+}
+
+func TestBuildFromGenesis_WhenGenesisDoesNotExists(t *testing.T) {
+	t.Parallel()
+
+	bs, err := BuildFromGenesis("/not/exists/genesis.json", 0)
+	require.Nil(t, bs)
+	require.Error(t, err, os.ErrNotExist)
+}
+
+func TestWriteGenesisSpecFileWhenFileAlreadyExists(t *testing.T) {
+	t.Parallel()
+
+	f, err := ioutil.TempFile("", "existing file data")
+	require.NoError(t, err)
+	defer os.Remove(f.Name())
+
+	someBytes := []byte("Testing some bytes")
+	err = WriteGenesisSpecFile(someBytes, f.Name())
+
+	require.EqualError(t, err,
+		fmt.Sprintf("file %s already exists, rename to avoid overwriting", f.Name()))
+}
+
+func TestWriteGenesisSpecFile(t *testing.T) {
+	t.Parallel()
+
+	cfg := NewTestConfig(t)
+	cfg.Init.Genesis = "../chain/gssmr/genesis.json"
+
+	expected, err := genesis.NewGenesisFromJSONRaw(cfg.Init.Genesis)
+	require.NoError(t, err)
+
+	err = InitNode(cfg)
+	require.NoError(t, err)
+
+	bs, err := BuildFromGenesis(cfg.Init.Genesis, 0)
+	require.NoError(t, err)
+
+	data, err := bs.ToJSONRaw()
+	require.NoError(t, err)
+
+	tmpFiles := []string{
+		t.TempDir() + "/unique-raw-genesis.json",
+	}
+
+	for _, tmpFile := range tmpFiles {
+		err = WriteGenesisSpecFile(data, tmpFile)
+		require.NoError(t, err)
+		require.FileExists(t, tmpFile)
+
+		defer os.Remove(tmpFile)
+
+		file, err := os.Open(tmpFile)
+		require.NoError(t, err)
+		defer file.Close()
+
+		gen := new(genesis.Genesis)
+
+		decoder := json.NewDecoder(file)
+		err = decoder.Decode(gen)
+		require.NoError(t, err)
+
+		require.Equal(t, expected.ChainType, gen.ChainType)
+		require.Equal(t, expected.Properties, gen.Properties)
+	}
+}
+
+func TestBuildFromDB(t *testing.T) {
+	t.Parallel()
+
+	// setup expected
+	cfg := NewTestConfig(t)
+	cfg.Init.Genesis = "../chain/gssmr/genesis.json"
+	expected, err := genesis.NewGenesisFromJSONRaw(cfg.Init.Genesis)
+	require.NoError(t, err)
+	// initialise node (initialise state database and load genesis data)
+	err = InitNode(cfg)
+	require.NoError(t, err)
+
+	bs, err := BuildFromDB(cfg.Global.BasePath)
+	require.NoError(t, err)
+	res, err := bs.ToJSON()
+	require.NoError(t, err)
+	jGen := genesis.Genesis{}
+	err = json.Unmarshal(res, &jGen)
+	require.NoError(t, err)
+
+	require.Equal(t, expected.Genesis.Raw["top"]["0x3a636f6465"], jGen.Genesis.Runtime["system"]["code"])
```

Comment:
I have no clue what's going on here, but if `Raw["top"]["0x3a636f6465"]` is meant to be `["system"]["code"]` then be it! 😄

Reply:
This is confirming that the `code` that is the generated file is the same as what was included in the original genesis file.

Reply:
can we add `const codeHex = "0x3a636f6465"` and use this var in there?

Reply:
Also if these constants are not used anywhere else, maybe define them in-lined where they're required? Locally scoping 💯